contentcuration/contentcuration/utils/appnexus/base.py

```diff
@@ -0,0 +1,81 @@
+from abc import ABC, abstractmethod
+from builtins import NotImplementedError
+from typing import Optional, Union, Dict
+
+
+class Backend(ABC):
+    """ An abstract base class for backend interfaces that also implements the singleton pattern """
+    _instance = None
+
+    def __new__(class_, *args, **kwargs):
+        if not isinstance(class_._instance, class_):
+            class_._instance = object.__new__(class_, *args, **kwargs)
+        return class_._instance
+
+    @abstractmethod
+    def connect(self) -> None:
+        """ Establishes a connection to the backend service. """
+        pass
+
+    @abstractmethod
+    def make_request(self, url: str, params=None) -> Union[bytes, str, Dict]:
+        """ Makes an HTTP request to a given URL using the specified method. """
+        pass
+
+    @abstractmethod
+    def request(self) -> None:
+        """ Blueprint for the request object. """
+        pass
+
+    @abstractmethod
+    def response(self) -> None:
+        """ Blueprint for the response object. """
+        pass
+
+    @classmethod
+    def get_instance(cls) -> 'Backend':
+        """ Returns existing instance, if not then create one. """
+        return cls._instance if cls._instance else cls._create_instance()
+
+    @classmethod
+    def _create_instance(cls) -> 'Backend':
+        """ Returns the instance after creating it. """
+        raise NotImplementedError("Subclasses should implement the creation of instance")
+
+
+class BackendFactory:
+    def create_backend(self, backend: Optional[Backend] = None) -> Backend:
+        """
+        Create a Backend instance based on Django or manual settings.
+
+        Args:
+            backend (Optional[Backend], optional): 
+            An optional pre-existing Backend instance.
+
+        Returns:
+            Backend: A Backend instance.
```

Comment:
I'm not an expert in python docstrings, but this looks like an alternative format compared with the format we use already in the codebase